File: linkedlist/25/my.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        # Edge case: empty list or k=1 (no reversal needed)
        if not head or k == 1:
            return head
        
        # Create dummy node
        dummy = ListNode(0)
        dummy.next = head
        
        # Initialize pointers
        prev_group_end = dummy
        current = head
        
        # Function to reverse k nodes starting from start
        def reverse_k_nodes(start, k):
            # Initialize pointers for reversal
            prev, curr = None, start
            
            # Reverse k nodes
            for _ in range(k):
                next_temp = curr.next
                curr.next = prev
                prev = curr
                curr = next_temp
            
            # Return new start (prev) and the next node after group (curr)
            return prev, curr, start
        
        # Count remaining nodes
        count = 0
        node = head
        while node:
            count += 1
            node = node.next
        
        # Process full k-groups only
        while count >= k:
            # Get the current group to reverse
            group_start = current
            
            # Reverse current k-group
            new_group_start, next_group_start, new_group_end = reverse_k_nodes(group_start, k)
            
            # Connect previous group end to new group start
            prev_group_end.next = new_group_start
            
            # Connect new group end to next group start
            new_group_end.next = next_group_start
            
            # Update pointers for next iteration
            prev_group_end = new_group_end
            current = next_group_start
            
            # Update remaining count
            count -= k
        
        return dummy.next

File: recursion/25/my.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

 struct ListNode* reverseKGroup(struct ListNode* head, int k) {
    // Edge cases
    if (head == NULL || k == 1) {
        return head;
    }
    
    // Create a dummy node
    struct ListNode dummy;
    dummy.next = head;
    
    // Count total nodes
    int count = 0;
    struct ListNode* counter = head;
    while (counter != NULL) {
        count++;
        counter = counter->next;
    }
    
    // Initialize pointers
    struct ListNode* prevGroupEnd = &dummy;
    struct ListNode* curr = head;
    
    // Process complete k-groups
    while (count >= k) {
        // Save the start of the group (will be the end after reversal)
        struct ListNode* groupStart = curr;
        
        // Find the start of the next group
        struct ListNode* nextGroupStart = curr;
        for (int i = 0; i < k; i++) {
            nextGroupStart = nextGroupStart->next;
        }
        
        // Reverse the current k-group
        struct ListNode* prev = nextGroupStart; // Next group becomes previous of reversed group
        for (int i = 0; i < k; i++) {
            struct ListNode* next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        
        // Connect the previous group to the reversed group
        prevGroupEnd->next = prev;
        
        // Update pointers for next iteration
        prevGroupEnd = groupStart;
        
        // We've processed k nodes
        count -= k;
    }
    
    return dummy.next;
}